docs/DafnyRef/UserGuide.md

```diff
@@ -1402,7 +1402,7 @@ We can also classify the assertions extracted by Dafny in a few categories:
 
 * Every value whose type is assigned to a [subset type](#sec-subset-types) yields an _assertion_ that it satisfies the subset type constraint.
 * Every non-empty [subset type](#sec-subset-types) yields an _assertion_ that its witness satisfies the constraint.
-* Every [Assign-such-that operator](#sec-update-and-call-statement) `x :| P(x)` yields an _assertion_ that `exists x :: P(x)`.
+* Every [Assign-such-that operator](#sec-update-and-call-statement) `x :| P(x)` yields an _assertion_ that `exists x :: P(x)`. To be compilable, the value of `x` must also be uniquely determined, yielding an additional _assertion_ that `forall x, y | P(x) && P(y) :: x == y`.
```

Comment:
This isn't technically what we do. Maybe something like this?
```suggestion
* Every [Assign-such-that operator](#sec-update-and-call-statement) `x :| P(x); Body(x)` yields an _assertion_ that `exists x :: P(x)`. In case `x` is non-ghost, it also yields `forall x, y | P(x) && P(y) :: Body(x) == Body(y)`.
```

Reply:
Wow ... this is surprising to me ... in various ways.

I didn't know that the `Body` was checked. It feels weird to me, but it has a certain logic to it: It ensures that the function is deterministic, even though the assign-such-that may not be :slightly_smiling_face: 

The reference to `Body` only makes sense in functions. So I checked what happens in methods and found - nothing! In methods you don't need to prove determinism:

```dfy
method Bar() returns (result: int)
{
  assume exists x :: P(x);
  var x :| P(x);
  result := x;
}
```
verifies :astonished:

Why do non-ghost functions need to be more deterministic than non-ghost methods? In my head functions are more abstract while methods are more explicit, algorithmic, which would rather lead me to expect that non-determinism in functions could be fine while it should be prohibited in methods :see_no_evil: 

Though, searching the documentation for "deterministic" does make feel that I should have known this: Already the first three findings clearly state that functions are pure, deterministic while methods may be non-deterministic :confused: 
I understand that functions often need to satisfy stricter criteria to ensure that inconsistencies are impossible, but why does this have to extend to determinism in non-ghost functions?

Anyhow, I think that the function/method distinction needs to be mentioned. What do you think about

>* Every [Assign-such-that operator](#sec-update-and-call-statement) `x :| P(x)` yields an _assertion_ that `exists x :: P(x)`. In case the Assign-such-that operator `x :| P(x); Body(x)` appears in a function and `x` is non-ghost, it also yields `forall x, y | P(x) && P(y) :: Body(x) == Body(y)`.

Reply:
Wow it blows my mind too. I always supposed that the value of x must be uniquely determined, but currently it's the value of the let-such-that expression indeed. So whereas the following verifies:
```dafny
function test(): int {
  var x :| 0 < x < 3;
  1
}
```

the following does not

```dafny
function test(): int {
  var x :| 0 < x < 3;
  x + 1
}
```

In either cases, really, @RustanLeino proves me otherwise,
I think we should not need the unicity in functions anyways. Unicity should only be a nice to have.
Indeed , we can view a let such that assignment
```
var x :| P(x); Body(x)
```

as an assignment from a call to a uniquely chosen deterministic function
```
var x := choose<UniqueNumber>(x => P(x)); Body(x)
```
that is invisible to the user and cannot be referenced.
The fact that Body(x) will always have the same value is useful if we want to make a proof about that function and create another let such that assignment, but it does not have to be a requirement.

Meantime, I think that your PR looks good.

Reply:
Done: https://github.com/dafny-lang/dafny/compare/a3149599fbe72cb8547d978a0ae7aea8abaf2aa1..9df67ab939b38ba93d87f5d338ff4e46de0ad63a :)